### PR TITLE
core(config): clean flags for settings

### PR DESCRIPTION
**Pull request by patrickhulce**
alternative to #4958

this brings back -A, it didn't come out quite as clean as I had hoped, but it is typesafe which is kinda nice :)

### CODE REVIEW

File: typings/config.d.ts [hunk 1]
```diff
@@ -24,11 +24,11 @@ declare global {
         settings?: SettingsJson;
         passes?: PassJson[];
       }
-
```
Comment (patrickhulce) [inline typings/config.d.ts]: someone's got a diff trailing space editor setting it seems :)

File: lighthouse-core/config/config.js [hunk 2]
```diff
@@ -246,8 +246,24 @@ function expandArtifacts(artifacts) {
   return artifacts;
 }
 
+/**
+ * @param {LH.Flags=} flags
+ * @return {Partial<LH.Config.Settings>}
+ */
+function cleanFlagsForSettings(flags = {}) {
```
Comment (paulirish) [inline lighthouse-core/config/config.js]: semantically this is `removedUndefinedProperties` right?

could we call it that instead?

Comment (patrickhulce) [reply lighthouse-core/config/config.js]: not sure I'm following you, this is looking at what properties are not on `constants.defaultSettings` and dropping them from flags regardless of their value.

the JSON.parse(JSON.stringify()) approach would address the specific undefined yargs case but not an underlying problem that having miscellaneous junk present that isn't part of settings throws off our isDeepEqual later

Comment (paulirish) [reply lighthouse-core/config/config.js]: oh got it. 

>  what properties are not on constants.defaultSettings

ahhh.. mmhmmm. 

k lgtm

File: lighthouse-core/config/constants.js [hunk 3]
```diff
@@ -26,10 +26,24 @@ const throttling = {
   },
 };
 
+/** @type {LH.Config.Settings} */
```
Comment (brendankenny) [inline lighthouse-core/config/constants.js]: 👍

File: typings/config.d.ts (hunk 1 above)
Comment (brendankenny) [reply typings/config.d.ts]: ah, damn, eslint isn't checking .ts files, obvs. I'll find the setting :P

File: lighthouse-core/config/config.js [hunk 4]
```diff
@@ -246,8 +246,24 @@ function expandArtifacts(artifacts) {
   return artifacts;
 }
 
+/**
+ * @param {LH.Flags=} flags
```
Comment (brendankenny) [inline lighthouse-core/config/config.js]: maybe add a function description that this is dropping any flag not found on settings?

**Pull request merged**